src/cachemachine/rubintagfuncs.py

```diff
@@ -0,0 +1,259 @@
+            # mypy doesn't know that
+            _, nname, _, _ = parse_tag(rest)  # We only care about the name.
+            name = f"Experimental {nname}"
+    else:
+        # Everything else does get an actual semantic version
+        build = cycle_and_rest_to_optional_buildstring(cycle, ctag, rest)
+        typename = titlecase(tagtype.name)
+        restname = name[2:]
+        if (
+            tagtype == RubinTagType.RELEASE
+            or tagtype == RubinTagType.RELEASE_CANDIDATE
+        ):
+            # This is bulky because we don't want to raise an error here
+            # if we cannot extract a required field; instead we let the
+            # field be None, and then the semantic version construction fails
+            # later.  That's OK too, because we try that in a try/expect block
+            # and return None if we can't construct a version.  In *that*
+            # case we have a tag without semantic version information--which is
+            # allowable.
+            major = maybe_int(md.get("major"))
+            minor = maybe_int(md.get("minor"))
+            patch = maybe_int(md.get("patch", "0"))  # If omitted, it's zero
+            restname = f"r{major}.{minor}.{patch}"
+            pre = md.get("pre")
+            if pre:
+                pre = f"rc{pre}"
+                restname += f"-{pre}"
+        else:  # tagtype is weekly or daily
+            year = md.get("year")
+            month = md.get("month")
+            week = md.get("week")
+            day = md.get("day")
+            major = maybe_int(year)
+            if tagtype == RubinTagType.WEEKLY:
+                minor = maybe_int(week)
+                patch = 0
+                restname = f"{year}_{week}"  # preserve initial string format
+            else:
+                minor = maybe_int(md.get("month"))
+                patch = maybe_int(md.get("day"))
+                restname = f"{year}_{month}_{day}"  # preserve string format
+            pre = None
+        try:
+            semver = VersionInfo(
+                major=major,
+                minor=minor,
+                patch=patch,
+                prerelease=pre,
+                build=build,
+            )
+        except TypeError as exc:
+            logger.warning(f"Could not make semver from tag {tag}: {exc}")
+        name = f"{typename} {restname}"  # Glue together display name.
+        if cycle:
+            name += f"_{ctag}{cycle}"
+        if rest:
+            name += f"_{rest}"
+        cycle_int = maybe_int(cycle)
+    return (name, semver, cycle_int)
+
+
+def maybe_int(n: Optional[str]) -> Optional[int]:
+    if n is None:
+        return None
+    # We do the cast to float first to also catch the cycle case ('0019.001')
+    return int(float(n))
+
+
+def cycle_and_rest_to_optional_buildstring(
+    cycle: Optional[str],
+    ctag: Optional[str],  # if present, either 'c' or 'csal'
+    rest: Optional[str] = None,
+) -> Optional[str]:
```

Comment:
I guess it wouldn't hurt to add a comment to the effect "Any other stuff ends up as the 'build' field in the semantic version object; that is defined as a dot-separated alphanumeric string, so we massage our 'cycle' and 'rest' fields into something compatible with that."

Reply:
Oh, "buildstring" is actually the "build" component of the semver. SQR-059 has the context that I was missing. Would definitely be worth a reference to it here, and I would also recommend avoiding the phrase "buildstring" because in my experience that's used as a label for a build rather than for the last component of a semver version. I think what this is doing is something more like `construct_semver_build_component`.

Reply:
done